compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1192,6 +1192,8 @@ fn add_sanitizer_libraries(sess: &Session, crate_type: CrateType, linker: &mut d
     // are currently distributed as static libraries which should be linked to
     // executables only.
     let needs_runtime = !sess.target.is_like_android
+        && (!sess.opts.cg.link_self_contained.is_sanitizers_disabled()
+            || sess.opts.cg.link_self_contained.is_sanitizers_enabled())
```

Comment:
This condition is getting a bit too unreadable, I think. Can we split it into multiple guard clauses, each with a separate comment?

```rust
if sess.target.is_like_android {
    // sanitizer runtime libraries are provided dynamically
    // on Android targets.
    return; 
}

if /* disabled via link_self_contained */ {
    // ...
    return;
}

if /* disabled for crate type */ { 
   // ...
   return
}
```

Reply:
Also, it looks like this should also take the settings from the target spec into account (see `fn self_contained_components` in this file). Maybe that could also be used to remove the special case for Android?

Although, it looks to me like this check against what the target allows should happen much earlier in the process and not shortly before the linker is invoked. @lqd, what do you think (since this was added by https://github.com/rust-lang/rust/pull/116035)? Or maybe I'm overlooking something?

Reply:
It's likely we could move some of the checks earlier in some cases, but other checks/features do depend on the inferred linker flavor and thus couldn't all be moved that much earlier (IIRC sanitizers support is not long after the flavor has been inferred, and does depend on it a bit). Some other components can themselves also be inferred -- e.g. some variants in `LinkSelfContainedDefault` -- and also currently happen quite "late".

I don't myself have yet a good clear vision on how to structure all these concerns in the future, whether by splitting the currently-tightly-grouped checks into different early and late groups, or something else, but maybe @petrochenkov does?

Reply:
I'm happy to make any requested changes.